### DISCUSSION TITLE
CtrlrX v5.6.28 Release - Scalable GUI for responsive design added and menuBar not showing on export FIXED

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#588] [General]
Hi, 
I just published a new commit to my Ctrlr Fork as v5.6.28.
https://github.com/damiensellier/CtrlrX

There's 2 major updates in this commit.
1. The menu bar not showing up on exports is now fixed for v5.6.23+ and can be set in the property pane.
2. I added a callback triggered when the app/plugin frame is resized. It allows to have a LUA script and resize the uiZoom value to change relative to the Native OS window size.

![responsive exemple](https://github.com/RomanKubiak/ctrlr/assets/55164385/8640c9e1-0794-4170-9373-d1feb376a553)

This is the new Editor part for Globals in the property pane:

![5.6.28 upate](https://github.com/RomanKubiak/ctrlr/assets/55164385/7ee22eb7-8301-481a-a125-2f566212c463)

"**Canvas Bounds**" , changing values are now updating the actual canvas graphics. Before, only resizing with the mouse worked.
"**ViewPort is resizable on a restricted instance**" is set to to Resizable as default. You can lock the window with this option.
"**ViewPort Scrollbars visibility on an exported instance**" is set to "Show" as default". You can hide them to prevent scrollbars showing if the APP window is showing scrollbars on your panel when locked.
"**ViewPort resize limits enabled on an exported instance**" it's set to Disabled by default. It let you define a minimum and maximum APP window size.
"**ViewPort enable aspect ratio**" is where the magic happens. It's set to Disabled by default. When Enabled, you can keep an aspect ratio for your APP window, required for scalable/responsive design.
"**ViewPort fixed aspect ratio**" is calculated automatically when the canvas size field is changed (0 0 600 400 for 1.5 etc)
"**Panel ViewPort background colour**" is a new property to define the color of the ViewPort (where the panel canvas sits)

![5.3.28 VP cbk on resize dropdown](https://github.com/RomanKubiak/ctrlr/assets/55164385/3c9dcab5-dddb-4117-94d2-ecb2e529fc37)

"**Called when the ViewPort is resized**" is where you select your LUA script triggered by resizing the APP/Plugin window.

Here is an exemple on how to scale the panel to the OS window:

```
--
-- Called when the panelEditor is resized
--

vpResizedCbk = function(--[[ CtrlrPanelEditor --]] panel)

canvasWidth = panel:getCanvas():getWidth()
canvasHeight = panel:getCanvas():getHeight()
viewportWidth = panel:getPanelEditor():getWidth()
viewportHeight = panel:getPanelEditor():getHeight()
canvasRatio = canvasWidth/canvasHeight
zoomWidthRatio = viewportWidth/canvasWidth
zoomHeightRatio = viewportHeight/canvasHeight

if canvasWidth >= canvasHeight then
    zoomRatio = zoomWidthRatio
else
    zoomRatio = zoomHeightRatio
end

panel:getPanelEditor():setProperty("uiPanelZoom",zoomRatio,true)

end
```

Hope it will help with your new projects and happy new year 2024 to each and everyone of you ;)

Damien

EDIT : 

Saving/recalling the window sizes in a value tree stateData LUA script is required to have the current size stored in the plugin host prefs.
Exemple here (not responsive yet, but saving size for DAW) :
https://github.com/RomanKubiak/ctrlr/files/7819295/Select.Window.Size_1_0_0_2022-01-05_21-15.panel.zip
I'll post a sample panel with all features soon.

**Comment by Ted-juh** [comment]
Some great additions there Damien.  I haven't yet had time to test it, but I have to ask. Does this function "blur" pictures used for sliders/ backgrounds as long as they're not vector-based? Nonetheless, they are great additions. I'm in the process of writing a complete background for my panel which consists solely of the paint method, which would scale nicely with your added zoom function because it's based on the width/ height of the panel instead of absolute coordinates.

In my instance of Ctrlr, I added a few missing defs from classes. One that comes to mind is jmap. I also tried to bind the Normalisable Range class from Juce with Luabind. Which partially succeeded. Very handy when it comes to creating sliders yourself in a custom component. Yes, there are workarounds, like adding a function in Lua but it's fewer scripts and functions this way.

Not added yet but I also wanted to add a right-click menu to the custom component like the other components have. Speaking about right-click.  Wanted to add the "open panel mode" to the right-click menu on components, as well as a window that opens just the component properties in a floating window. 

I would like to add them to your fork so you can review them before adding them but.. I don't get the geist of GitHub. I can create my own fork but I don't have a clue what I'm doing when it comes to pulling/ requests/ commits and so on. Let alone contribute to your fork. Dnaldoog once tried to pull me through but to no avail. Read a lot but there are way too many confusing tutorials on how to do it in Visual Studio or GitHub's own program or external consoles. I just can't get the hang of it. 

So if someone has a tutorial that made it tick the "gotcha" boxes, please let me know.

**Reply by damiensellier** [reply]
Hi, @Ted-juh, yes pictures will be upscaled then "blurred". BUT, you can export your design x2 or x4 of the size and stretch to fit to a 1/4 sized component to get hi quality images even in 4k full screen without artefacts, I did that to some of my panels to have hi res on retina dispays. for a 50x50pixel button, just export a 200x200px and put the image to a 50x50 component. You'll get full resolution in uiZoom x4.

Right click is a nice idea it's like we did for the mouseExit, it's just a matter of adding a new callback, now we know how to do it :) I'll add it to my to-do list for future versions of my fork.

Forget about pulling request to the Main release of CTRLR because the creator is no longer active and will never patch the original version.
Concerning github, forks, commits etc, I highly suggest to use github desktop, it's an app for macOS or WIndows and do everything needed without having to use the terminal commands for syncing.
There's a good tutorial about github desktop and its different functions there : 
https://www.youtube.com/watch?v=MaqVvXv6zrU

When you do updates to the code on your fork, do little by little for every new feature added like "right click added to sliders" on a specific "branch". This way everyone can cherry pick the updates to their own forks. And it also has to be 100% tried and tested before commiting of course.

I'll post  pre-release builds for macOS and Windows of my version of 5.6.28 very soon. It has all the other updates I found useful from other forks. It will help all the users not able to build apps from source code and just download and enjoy.

My next work will be to fix the annoying plugin ID and naming when exporting a VST3 https://github.com/RomanKubiak/ctrlr/discussions/394. Once I have it solved, I'll do a release on my fork as CtrlrX v6.0 with everything available for anyone.

Happy new year 2024!

**Reply by Ted-juh** [reply]
I get what you're saying about the high quality images. Neat trick, never thought of doing it that way. 

But since I'm getting quite adapted to the custom component and the options provided with the graphics class, I kind of see it as a challenge to not use any pictures at all. 

But about GitHub. I was asking about how to add to your fork of Ctrlr. Is there a way? Do you want that? Or is it each on its own? Or do I maintain my own fork (?) of which you pick the things you might find usefull? 

I remember I made a Ctrlr panel that output a file to change the Plugin-ID which you had to import before compiling the vst3. But it should be possible to take the name/ creator of the panel from the ones you fill in when starting a new panel. The code for the conversion to the Plugin ID is already in Juce. Good luck with that.. and if you need any help (although my c++ skills sucks big time) just tell.

**Reply by damiensellier** [reply]
@Ted-juh maybe yu could try converting your design to svg or vector paths?
I'm into the vst3 naming problem I'll add something in the juce vst3 wrapper to grab the name and other IDs from the info.plist and windows program header, exactly as it's implemented as in the vst2wrapper, AU wrapper and AAX wrapper. 
The way the VST3 wrapper is done is totally enclosed and doesn't communicate with the Ctrlr manager at any point. So the only way to identify the vst3 plugin/panel is to get those IDs from a xml or datas available outside. We need those IDs at the entry point of the wrapper, before the wrapper even calls the ctrlr application. The vst2 wrapper doesn't even call the getName() in the audioprocessor class, it just gets the name from the plugin file itself (without extension), if you rename it, the plugin will show up differently in the host. VST3 is a bit different and creates the wrapper with statics from the projucer definitions, this is the reason why  it's stays "Ctrlr | Instigator" because it's hard-coded to the wrapper that is cloned when exporting the panel as vst3. There's no interaction with the panel possible in anyway as it is right now. It's totally segregated. Getting IDs from the header or the package bundle is the easiest with the Juce File class and XMLelement class.
 

I did the part of storing those IDs in the info.plist and windows program header already, so now I have them available at hand when exporting.
As you can see, now we have everything required, the panel version , the panel name, panel ID, manufacturer and manufacturer ID, so it's ready to go for the wrapper.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
  ///

    <key>CFBundleName</key>
    <string>exportNew01</string>  // for plugin name
///
    <key>CFBundleShortVersionString</key>
    <string>2.1</string>
    <key>CFBundleSignature</key>
    <string>5PN2</string> // for plugin UID
///
    <key>CFBundleVersion</key>
    <string>2.1</string>
///
    <key>NSHumanReadableCopyright</key>
    <string>damien</string> // For Manufacturer
    <key>ManufacturerID</key>
    <string>qmWq</string> // For ManufacturerID
  </dict>
</plist>

```

 The remaining part is to get them and pass them to the host, I'm half way through. Unfortunately I'm pretty busy from now on after holidays and will have to wait until february to tackle this issue once and for all. 

Concerning the github, you can probably request a push to my fork, I'll get a notification and I'll cherry pick the updated code for a new commit on a new branch of mine. I only got the updates from other forks by doing a pull request to my fork.

**Reply by damiensellier** [reply]
Edit: Yeah and you can also share your updates here in a discussion presenting the new feature with a link to your branch ;-) That would be so nice to have an updated version of Ctrlr including all the features everyone worked on.
Take care
Damien

**Reply by bijlevel** [reply]
Very good initiative, much appreciated! So now it's 'feature request' time 😉! I support @Ted-juh request for the right clicks. Maybe also a keyboard shortcut for the Lua editor. Most of the time the editor is hidden by the panel (I'm working on a rather big panel), so to show the editor a shortcut would be handy. Maybe I can do it myself, but if one of you could point me in the right direction, that would be nice! I'm also always struggling with the inspector panel layout. Never can I find quickly a property or setting. I'm not sure if rearranging is feasible. Maybe just colouring would help. I don't now. Again, if you can point me in the right direction where I have to look for these kind of changes, that would be very welcome.

**Reply by damiensellier** [reply]
@bijlevel I just checked the code and right click is already available on the customComponent. It's just that you have to right click on the name of the component to display it, not the empty area.
<img width="430" alt="Capture d’écran 2024-01-11 à 07 08 59" src="https://github.com/RomanKubiak/ctrlr/assets/55164385/3abfcac0-c7a6-4e39-b896-6dffb2a336fc">

Or are you talking about a right click callback event to trigger a lua script?

**Reply by damiensellier** [reply]
everything happening on the canvas area is related to CtrlrPanelCanvasHandlers.cpp and CtrlrPanelCanvas.cpp
It's going to be very easy to add functions in the popup like "display component properties" etc
I'll add this to my to do list. I'm thinking about creating a thread for that like a ctrlr new featuress request and bug fix.
I will definitely implement the look and feel class to be able to style the components with the stock JUCE color templates. That would be neat.

**Comment by Ted-juh** [comment]
> Maybe also a keyboard shortcut for the Lua editor. 
There is already a shortcut, it's "Ctrl + e". I guess it's "Command + e" on Mac.

> I'm not sure if rearranging is feasible. 
Rearranging is possible. Or add a section to set all the colors for background, text, button on, button off and so on,  that are now positioned between the options. The Component section and the Component Colours section.

>Maybe just coloring would help. 
It would help but the LookAndFeel Class for the PropertyPanel doesn't provide options to colour sections. Well, not that I could find.

I botched my installation of Ctrlr that I worked on to debug and release. I don't know what happened but I tried to fork Ctrlr (succeeded) but now I'm getting build errors while I did the exact same steps as I did before from Bijlevel's manual.

I think the problem is that it's not taking the repository from April 8th 2021. That was the one that would debug and release fine. But the 2 newer commits didn't. Does anybody know how I point Visual Studio to the right date of the repository?

**Reply by Ted-juh** [reply]
@ Damien

Even if the custom component is filled with graphics, I thought it was a transparency issue not selecting the component to register a right-click, then there's still no right-click menu. I could set the name to invisible later on but it's weird behaviour.

**Reply by bijlevel** [reply]
@Ted-juh @damiensellier I can confirm this behaviour on a Mac. Only r-click on the name or the border does show the menu.

**Reply by bijlevel** [reply]
The shortcut 'command+e" does set the panel in panel mode, so it does not show the Lua editor, at least in the Mac version. Of course it's a minor issue, but it would cause less mouse travel if there is a shortcut.

BTW my name is Sjoerd, so no need anymore to use 'bijlevel'😉.

**Comment by damiensellier** [comment]
Saving/recalling the window sizes in a value tree **stateData** LUA script is required to have the current size stored in the plugin host prefs.
Exemple here (not responsive yet, but saving size for DAW) : 
https://github.com/RomanKubiak/ctrlr/files/7819295/Select.Window.Size_1_0_0_2022-01-05_21-15.panel.zip
I'll post a sample panel with all features soon.

**Comment by dobo365** [comment]
Tested and working fine. Thx a lot @damiensellier 
Almost perfect because the zoom is not perfect.
If you align the width, at a smaller zoom, one can see the canvas is not big enough vertically. At bigger zoom, the canvas is bigger and is visible.
What can the reason be? 
Is it internal in CtrlrX? Is the Viewport fixed aspect ratio stored as double?
In my case, in the property I see 1,9011 while the exact value is 1,901084 (which is in fact the canvas aspect ratio 1403/738).

I don't know how the calculation is done within CtrlrX. Will check that but it should be:
- take window width
- use the fixed canvas aspect ratio to calculate the window height based on the window width and taking into account the top header bar of the window

1:1 Width= 1403
<img width="1363" height="747" alt="image" src="https://github.com/user-attachments/assets/e5153d50-4154-4ad1-87ed-2ab26f264a47" />

Bigger (small part of canvas visible at bottom) Width= 1775
<img width="1775" height="972" alt="image" src="https://github.com/user-attachments/assets/d12b9df2-411b-4610-b3d1-447f428e4e55" />

Smaller (canvas is smaller and thus synth no looking as 1:1) Width= 692
<img width="692" height="380" alt="image" src="https://github.com/user-attachments/assets/79408a1d-0145-4683-a8ed-1b42d58a7f8d" />

**Reply by damiensellier** [reply]
You can develop the callback method a little bit better to round up/down the width height values and limit the decimal space for the zoom or aspect ratio if needed. I never had a problem in my case because I almos always use 1.5 ratio or 2x 

Another trick to hide the canevas line is to set the viewport background colour to the same as your gui, this way you'll never see that thin grey bottom line.

**Reply by damiensellier** [reply]
Note, if you really want to improve the GUI at higher zoom ratio, you need to use SVG vector elements, as we do for responsive design for web. you can also design your elements with X2 or X3 or even x4 size of the zoom x1 size. Let's say your buttons are 50x20px export them full scale at 100x40 or 150x60px. This way you'll always have full resolution for your graphics.

Good luck!

**Reply by dobo365** [reply]
(of course) I have a version with the same background color. I made it light grey to show you the issue.
But I'm wondering what the issue is in fact. I don't think it is at the level of the callback method.
I mean, we have a canvas ratio which is fixed (and in fact whatever the number of decimals). 
The window width is changing and we calculate the zoom based on the canvas width.
The height should just change accordingly in the same proportion.
Seems that the zoomed canvas vertical size is becoming bigger or smaller than its expected value based on the fixed ratio.
Where are all those things calculated? 
CtrlrPanelCanvas.cpp?
CtrlrViewport.cpp?

Components/elements wise, the zoom works fine, no need to make things more complex.

**Reply by damiensellier** [reply]
It's happening there for the plugins : https://github.com/damiensellier/CtrlrX/blob/afcadd4ea769dbdb11aa37ff0d0906330dc9fb7a/Source/UIComponents/CtrlrApplicationWindow/CtrlrEditor.cpp#L100

For the standalone app:

https://github.com/damiensellier/CtrlrX/blob/1ffa0a27ecaf5fd45fd57ed64df9981ae5387bd8/Source/Core/StandaloneWrapper/CtrlrStandaloneWindow.cpp#L94


https://github.com/damiensellier/CtrlrX/blob/1ffa0a27ecaf5fd45fd57ed64df9981ae5387bd8/Source/Core/StandaloneWrapper/CtrlrStandaloneWindow.cpp#L122


It's with a juce function setFixedAspectRatio()

The aspect ratio is based on the width of the container & its height INCLUDING the title bar which distorts the calculation.

You need to take a look at my post on the JUCE forum about that, I did not double check on windows so it may be possible it's not working the same way. 
Look for vpOsWindowHeight in the code

https://forum.juce.com/t/setfixedaspectratio-and-problem-with-standalone-app-because-of-native-title-bar-height/59368

Maybe the plugin container also requires using the calculation taking in account the title bar height with getPeer() idk

**Reply by damiensellier** [reply]
Edit: 
I just checked and everything seems to be fine to me.
I don't know why the aspect ratio is wrong.

Could you run your plugin and do some screen capture for 3 sizes. Then measure in pixels the height and width of each one. You'd also need to check the height and width including the title bar and the borders left right and bottom.

You should get your correct aspect ratio either with the borders & title bar, or for the content only.

Let me know.

It would be nice also to check on PC John @dnaldoog instances of its latest Juno alpha software and if there's the same phenomenon.

**Reply by damiensellier** [reply]
Also please confirm if it's on the standalone version, or plugin version, or both ;)

**Reply by dobo365** [reply]
At this stage it is only for the standalone and I see the same behavior both on Mac M1 and on PC Windows.
The window height calculation is not well done, indeed due to the top window border which is bigger.

In fact we must enforce the original canvas aspect ratio which is also the value stored in the Viewport fixed aspect ratio property.
But you can recalculate it as panelCanvasWidth/panelCanvasHeight.
This will define the interior size of the window.
From there, one can have any border sizes.

I understand that setFixedAspectRatio sets the aspect ratio of the full new window. Correct?

**We need to calculate it in a different way only based on the Window internal width.**

FixedAR = panelCanvasWidth/panelCanvasHeight = windowInternalWidth/windowInternalHeight
NewAR to be calculated 
= vpOsWindowWidth/vpOsWindowHeight 
= vpOsWindowWidth / (vpOsFrameTop + vpOsFrameBtm + windowInternalHeight)
= vpOsWindowWidth / (vpOsFrameTop + vpOsFrameBtm + windowInternalWidth/FixedAR)
= vpOsWindowWidth / (vpOsFrameTop + vpOsFrameBtm + (vpOsWindowWidth - vpOsFrameLeft - vpOsFrameRight) / FixedAR)

I will try it...

**Comment by dobo365** [comment]
Tried on my M1 with
`    vpStandaloneAspectRatio = double(vpOsWindowWidth / (vpOsFrameTop + vpOsFrameBtm + (vpOsWindowWidth - vpOsFrameLeft - vpOsFrameRight) / (panelCanvasWidth/panelCanvasHeight))) ;`
in CtrlrStandaloneWindow.cpp but doesn't work. The aspect ratio is not kept. Need to understand why...
Will try tomorrow on my PC...
Unless you have an idea ;-)

**Reply by damiensellier** [reply]
That's strange because the calculation takes the os title bar and borders via getPeer() which should return the actual values whatever the os is.
I came with the current formula and tested it on my MacBook and mac mini and if I remember I also tested it on win 10 with no problem, it was a long time ago. 

This should need to be tested in debug mode and output the os specific borders and title bar values to the debug output or file to check if it corresponds to the actual size of we screen capture it. If it returns null this is probably where the discrepency is coming from.

Can you open an issue on CtrlrX about that? We'll try to make it work for every platform.

Thanks

**Reply by damiensellier** [reply]
If you can ask Gemini to add some _DBG() macros to return the os container border values that will be an easy one.

**Reply by dobo365** [reply]
I'm sorry Damien I have not any clue on how to use the debugger (VS or Xcode) for CtrlrX while running the app...
Is it needed to build for Debug? On my M1 I'm getting Ctrlr_SharedCode (standalone_plugin) is not testable.
Searched on Google about _DBG() and didn't find good explanation.

Will need your help (can you point me to a good explanation on how to use the debug mode in Xcode/VS?)
I know how a debugger is working, just need to know how to get it working while running my Ctrlr standalone panel...

**Reply by damiensellier** [reply]
Hi, _DBG() is a macro linked to Juce debug system.
It will output informations such as variables in the debug area of Xcode or vs when called. It's the same thing as console() in Lua from ctrlr.

When compiling a standalone exe with vs in debug mode, once the process is finished, the app will be running and attached to vs. The exe will then output those debugger codes everytime you hit a _DBG() functions. 

_DBG() are not going to compile in release mode.

Exemple : _DBG("This is a test you hit this particular line in your code.")

When you step into this line you should see "This is a test you hit this particular line in your code." In your vs debug area.

You can also pass some variables with concatenation with << _your_variable_ << 
Etc

Gemini will help you with that. Ask it to add debug outputs to a JUCE based plugin code it will add DBG() you'd need to precise your macro is _DBG() with the underscore instead and you'll get working debug outputs.

**Comment by damiensellier** [comment]
When you run an exe in debug mode but it's not been compiled and initiated by vs you'd need to attach this running app to vs ctrlrx project. You can search for "attach exe to vs for debugging" in Google. Same as explained above, once attached you'll see your DBG output in vs then.
